File: 338-counting-bits/338-counting-bits.java
class Solution {
    
    public int num_ones(int num) {
        int count = 0;
        
        // Set rightmost bit to 0 until num becomes 0
        while (num != 0) {
            count++;
            num = (num & (num - 1));
        }
        
        return count;
    }
    
    // Time = O(N)
    // Space = O(1)
    public int[] countBits(int n) {
        int[] res = new int[n+1];
        
        for (int i = 0; i < n + 1; i++) {
            res[i] = num_ones(i);
        }
        
        return res;
    }
}

File: 338-counting-bits/338-counting-bits.py
class Solution:
    # Time = O(n)
    # Space = O(1)
    def countBits(self, n: int) -> List[int]:
        table = [0] * (n+1)
        curr_offset = 1
        
        for i in range(1, n+1):
            # Update the curr_offset based on i value
            if i == curr_offset * 2: # i.e. powers of 2: 2, 4, 8, 16, ...
                curr_offset = curr_offset * 2
            
            table[i] = 1 + table[i-curr_offset]
        
        return table
        